airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py

```diff
@@ -3,37 +3,59 @@
 #
 
 import asyncio
+import itertools
 import logging
-from datetime import datetime
 from functools import cache
-from typing import Any, Iterable, List, Mapping, Optional, Union
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Union
 
-from airbyte_cdk.models import SyncMode
-from airbyte_cdk.sources.declarative.types import StreamSlice, StreamState
 from airbyte_cdk.sources.file_based.exceptions import MissingSchemaError, RecordParseError, SchemaInferenceError
 from airbyte_cdk.sources.file_based.remote_file import RemoteFile
-from airbyte_cdk.sources.file_based.schema_helpers import merge_schemas, type_mapping_to_jsonschema
+from airbyte_cdk.sources.file_based.schema_helpers import merge_schemas
 from airbyte_cdk.sources.file_based.schema_validation_policies import record_passes_validation_policy
 from airbyte_cdk.sources.file_based.stream import AbstractFileBasedStream
+from airbyte_cdk.sources.file_based.stream.cursor import FileBasedCursor
+from airbyte_cdk.sources.file_based.types import StreamSlice
+from airbyte_cdk.sources.streams import IncrementalMixin
 from airbyte_cdk.sources.utils.record_helper import stream_data_to_airbyte_message
 
 
-class DefaultFileBasedStream(AbstractFileBasedStream):
+class DefaultFileBasedStream(AbstractFileBasedStream, IncrementalMixin):
+
     """
     The default file-based stream.
     """
 
+    ab_last_mod_col = "_ab_source_file_last_modified"
+    ab_file_name_col = "_ab_source_file_url"
+    airbyte_columns = [ab_last_mod_col, ab_file_name_col]
+
+    def __init__(self, cursor: FileBasedCursor, **kwargs):
+        super().__init__(**kwargs)
+        self._cursor = cursor
+
+    @property
+    def state(self) -> MutableMapping[str, Any]:
+        return self._cursor.get_state()
+
+    @state.setter
+    def state(self, value: MutableMapping[str, Any]):
+        """State setter, accept state serialized by state getter."""
+        self._cursor.set_initial_state(value)
+
     @property
     def primary_key(self) -> Optional[Union[str, List[str], List[List[str]]]]:
         return self.config.primary_key
 
-    def read_records(
-        self,
-        sync_mode: SyncMode,
-        cursor_field: List[str] = None,
-        stream_slice: Optional[StreamSlice] = None,
-        stream_state: Optional[StreamState] = None,
-    ) -> Iterable[Mapping[str, Any]]:
+    def compute_slices(self) -> Iterable[Optional[Mapping[str, Any]]]:
+        # Group all files by timestamps and return them as slices
+        files = self.list_files_for_this_sync()
```

Comment:
I believe this should be a list sorted by `(last_modified, uri)` (and then we can get rid of the `slices.sort` on line 54). I noticed that when I reordered the items in `multi_csv_same_timestamp_more_files_than_history_size_scenario` the test failed (synced the lexographically earlier file).

Reply:
good catch. fixed and updated the test to cover that case